implementations/rust/ockam/ockam_api/src/cloud/email_address.rs

```diff
@@ -0,0 +1,210 @@
+use minicbor::{Decode, Encode};
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::{Error, Result};
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use std::fmt::Display;
+
+/// Newtype for an email address
+/// It is backed by a String and implements a PartialEq instance
+/// based on a lowercase comparison.
+///
+/// Note that the SMTP RFC (https://www.rfc-editor.org/rfc/rfc5321.txt, section 2.4) specifies that
+/// we should not make a case insensitive comparison on the local part of the email address
+///
+/// However we currently receive lowercase email addresses from the Controller in `ProjectUserRole`,
+/// and we need to make a case insensitive comparison when comparing with an email address in
+/// `UserInfo`.
+#[derive(Debug, Clone, Eq, Ord, PartialOrd, Decode, Deserialize, Encode, Serialize)]
+#[cbor(transparent)]
+#[serde(transparent)]
+pub struct EmailAddress(#[n(0)] String);
+
+impl EmailAddress {
+    /// Create a new EmailAddress without parsing it
+    /// To be used with care!
+    pub fn new_unsafe(s: &str) -> EmailAddress {
+        EmailAddress(s.to_string())
+    }
+
+    /// Parse an email address using the email_address crate
+    pub fn parse(s: &str) -> Result<EmailAddress> {
+        // validate the email address using the same regex as the one used by the Controller
+        // https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address
+        let regex = Regex::new(r"^[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$").map_err(Self::incorrect_regex)?;
+        if regex.is_match(s) {
+            Ok(EmailAddress(s.to_string()))
+        } else {
+            Err(Error::new(
+                Origin::Api,
+                Kind::Invalid,
+                format!("{s} is not a valid email address"),
+            ))
+        }
+    }
+
+    fn incorrect_regex(e: regex::Error) -> Error {
+        Error::new(
+            Origin::Api,
+            Kind::Invalid,
+            format!("incorrect regular expression {e:?}"),
+        )
+    }
+}
+
+/// Lowercase comparison for email addresses
+impl PartialEq for EmailAddress {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_lowercase() == other.0.to_lowercase()
+    }
+}
+
+impl Default for EmailAddress {
```

Comment:
Is it useful? I would remove it. Not an obvious implementation

Reply:
It is required because another struct is using `EmailAddress` and deriving `Default`.

Reply:
Which one? Maybe it should not?

Reply:
This is done for `UserInfo` but I did not realize that it was only done for tests. So I also added a `cfg(test)` attribute there and maybe we will do a refactoring for tests to remove those default instances (I'd rather have arbitraries for tests).